### PR TITLE
clp-package: Add handling for IR extraction jobs to the query scheduler and workers.

### PR DESCRIPTION
**Pull request by haiqi96**
# Description
This PR adds a new ExtractIR job and supports it in the CLP package. The change is a preparation for LogViewer support.
The IR extraction job takes 3 arguments: original_file_id, msg_ix and ir_target_size. The query scheduler use the first two arguments to find the file split to decompress, and creates a single task to extract the IR into a local path specified in the clp-config.

The PR makes the following changes:
1. Adds a new job type, ExtractIR and support it in the query scheduler
2. Refactors query scheduler to let IR extraction job and search job share common code
3. Add a new worker task extract_ir_task. Refactor the search task to let share common code with the new task. 
4. Updated start-clp.py to let different type of workers have specific mounts and enviroment
5. Add IR extraction related configs into the clp-config.

# Validation performed

- Submitted an IR extraction job with a helper scripts, confirmed that job finishes successfully and an IR is extracted.
- Submitted search and aggregation jobs via both commandline and webui. Confirmed that there is no error.

### CODE REVIEW

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 1]
```diff
@@ -604,20 +619,28 @@ def generic_start_worker(
...
     ]
+    if worker_specific_mount:
+        necessary_mounts.extend(worker_specific_mount)
+
     for mount in necessary_mounts:
-        if mount:
-            container_start_cmd.append("--mount")
-            container_start_cmd.append(str(mount))
+        if not mount:
+            raise ValueError(f"Required mount configuration is empty: {necessary_mounts}")
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:641]: If possible we should print out which mount is empty.. but don't see an easy way with the current code

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 2]
```diff
@@ -590,6 +711,51 @@ async def handle_finished_search_job(
     del active_jobs[job_id]
 
 
+async def handle_finished_extract_ir_job(
+    db_conn, job: SearchJob, task_results: Optional[Any]
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:715]:
```suggestion
    db_conn, job: ExtractIrJob, task_results: Optional[Any]
```

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 3]
```diff
@@ -590,6 +711,51 @@ async def handle_finished_search_job(
...
+) -> None:
+    global active_jobs
+
+    job_id = job.id
+    new_job_status = QueryJobStatus.SUCCEEDED
+    num_task = len(task_results)
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:721]: num_tasks

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 4]
```diff
@@ -590,6 +711,51 @@ async def handle_finished_search_job(
...
+    if 1 != num_task:
+        logger.error(
+            f"Unexpected number of task under IR extraction job: {job_id}. "
+            f"expected 1, got {num_task}"
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:725]:
```suggestion
            f"Unexpected number of tasks for IR extraction job {job_id}. "
            f"Expected 1, got {num_tasks}."
```

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 5]
```diff
@@ -364,6 +427,30 @@ async def acquire_reducer_for_job(job: SearchJob):
     logger.info(f"Got reducer for job {job.id} at {reducer_host}:{reducer_port}")
 
 
+def dispatch_job_and_update_db(
+    db_conn,
+    new_job: QueryJob,
+    target_archives: List[str],
+    clp_metadata_db_conn_params: Dict[str, any],
+    results_cache_uri: str,
+    num_task: int,
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:436]: num_tasks

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 6]
```diff
@@ -590,6 +711,51 @@ async def handle_finished_search_job(
...
+        db_conn,
+        QUERY_JOBS_TABLE_NAME,
+        job_id,
+        new_job_status,
+        QueryJobStatus.RUNNING,
+        num_tasks_completed=num_task,
+        duration=(datetime.datetime.now() - job.start_time).total_seconds(),
+    ):
+        if new_job_status == QueryJobStatus.SUCCEEDED:
+            logger.info(f"Completed job {job_id}.")
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:753]:
```suggestion
            logger.info(f"Completed IR extraction job {job_id}.")
```

File: components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py [hunk 7]
```diff
@@ -590,6 +711,51 @@ async def handle_finished_search_job(
...
+        else:
+            logger.info(f"Completed job {job_id} with failing tasks.")
```
Comment (kirkrodrigues) [inline components/job-orchestration/job_orchestration/scheduler/query/query_scheduler.py:755]:
```suggestion
            logger.info(f"Completed IR extraction job {job_id} with failing tasks.")
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/job-orchestration/job_orchestration/executor/query/utils.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+from job_orchestration.scheduler.scheduler_data import QueryTaskResult, QueryTaskStatus
+
+
+def get_task_log_file_path(clp_logs_dir: Path, job_id: str, task_id: int) -> Path:
+    worker_logs_dir = clp_logs_dir / job_id
+    worker_logs_dir.mkdir(exist_ok=True, parents=True)
+    return worker_logs_dir / f"{task_id}-clo.log"
+
+
+def report_command_creation_failure(
```
Comment (haiqi96) [inline components/job-orchestration/job_orchestration/executor/query/utils.py:22]: making this to be an individual function might be an overkill
@kirkrodrigues any suggestions?

Comment (kirkrodrigues) [reply components/job-orchestration/job_orchestration/executor/query/utils.py:22]: It's fine

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

clp-package: Add handling for IR extraction jobs to the query scheduler and workers.

**Pull request merged**